src/bin/cargo/commands/clean.rs

```diff
@@ -17,29 +21,243 @@ pub fn cli() -> Command {
+            )
+            .hide(true)
+            .value_name("DURATION"),
+        )
+        .arg(
+            opt(
+                "max-download-age",
+                "Deletes any downloaded cache data that has not been used since then given age (unstable)",
+            )
+            .hide(true)
+            .value_name("DURATION"),
+        )
+
+        .arg(
+            opt(
+                "max-src-size",
+                "Deletes source cache files until the cache is under the given size (unstable)",
+            )
+            .hide(true)
+            .value_name("SIZE"),
+        )
+        .arg(
+            opt(
+                "max-crate-size",
+                "Deletes crate cache files until the cache is under the given size (unstable)",
+            )
+            .hide(true)
+            .value_name("SIZE"),
+        )
+        .arg(
+            opt("max-git-size",
+                "Deletes git dependency caches until the cache is under the given size (unstable")
+            .hide(true)
+            .value_name("SIZE"))
+        .arg(
+            opt(
+                "max-download-size",
+                "Deletes downloaded cache data until the cache is under the given size (unstable)",
+            )
+            .hide(true)
+            .value_name("DURATION"),
+        )
+
+        // These are unimplemented. Leaving here as a guide for how this is
+        // intended to evolve. These will likely change, this is just a sketch
+        // of ideas.
+        .arg(
+            opt(
+                "max-target-age",
+                "Deletes any build artifact files that have not been used since then given age (unstable) (UNIMPLEMENTED)",
+            )
+            .hide(true)
+            .value_name("DURATION"),
+        )
+        .arg(
+            // TODO: come up with something less wordy?
+            opt(
+                "max-shared-target-age",
+                "Deletes any shared build artifact files that have not been used since then given age (unstable) (UNIMPLEMENTED)",
+            )
+            .hide(true)
+            .value_name("DURATION"),
+        )
+        .arg(
+            opt(
+                "max-target-size",
+                "Deletes build artifact files until the cache is under the given size (unstable) (UNIMPLEMENTED)",
+            )
+            .hide(true)
+            .value_name("SIZE"),
+        )
+        .arg(
+            // TODO: come up with something less wordy?
+            opt(
+                "max-shared-target-size",
+                "Deletes shared build artifact files until the cache is under the given size (unstable) (UNIMPLEMENTED)",
+            )
+            .hide(true)
+            .value_name("DURATION"),
+        )
+
         .after_help("Run `cargo help clean` for more detailed information.\n")
 }
 
 pub fn exec(config: &mut Config, args: &ArgMatches) -> CliResult {
-    let ws = args.workspace(config)?;
+    let ws = args.workspace(config);
```

Comment:
A mantra that I feel should be followed here is "the hackiest hack is the best hack" or rather "when throwing something together, be surgical rather than over-generalize".

The CLI is a placeholder atm so I'd expect it to be very surgical in how its implemented.  The parts of the GC we are working on for now are global, so I would expect a switch in `exec` for global vs non-global mode and not to also dive down deeper into the layers in the clean op.  Because this is trying to jump ahead to target stuff, we are making touching a lot more code than is appropriate for where this effort is atm

Reply:
I'm not sure I understand what this comment is getting towards. Our general policy is that the command level is responsible for parsing the CLI flags, and then handing off all the logic to the ops level. Since there is some important logic for deciding what actions need to be taken based on how the options combine together, I think it is good to keep that away from the command level. For example, I wouldn't want it to be possible to pass in an invalid configuration to the `ops::clean()` function. If that logic for whether or not the workspace is loaded was moved to `clean::exec()`, it would have to be duplicated again to validate that the caller set up the options correctly.

Reply:
> I'm not sure I understand what this comment is getting towards.

For being a placeholer, this feels fairly intrusive.  Seems like we could switch between different `ops` calls within the CLI.

> For example, I wouldn't want it to be possible to pass in an invalid configuration to the ops::clean() function. If that logic for whether or not the workspace is loaded was moved to clean::exec(), it would have to be duplicated again to validate that the caller set up the options correctly.

If we made the modes mutually exclusive (rather than additive) such that two separate ops can be decided between, than we can have something in and more easily talk through what the CLI should be, even if it ends up in `cargo clean`.  For example, I was just looking at the details of the `exec` and `ops` to check-in on what I was saying and noticed a subtle interaction that I'd want to talk about for the CLI that could easily be overlooked in the current semi-placeholder, semi-proposal route which would become obvious in a PR that takes a pure-placeholder and turns it into a proposal.

I have to say that I also have a bias in this thread coming from me not liking this being in `cargo clean` so to me this feels more like throwaway code and too much time can be spent on the details of it despite that (and later removing it)

Reply:
Ah, so perhaps we had a misunderstanding here, or at least a disagreement. I'm not seeing this as a "placeholder". This is a proposed interface that is intended to work generally as it is, but that I expect to *evolve* the specifics. For example, perhaps individual flags will be removed (like being able to control src/git/index separately), or changed to be higher level (like "delete things that can be recreated without a network connection") or maybe there is only one extra flag (like `--gc`). But my intention is very much that there is a single command for controlling all caches, and that the CLI interface has some options to control which caches are cleaned (with some default if no flags are specified).

So for me, this is not throwaway code. This is how I intend it to work. The open part is just the specific details of flag names and how general or specific they are.

Reply:
At this time, I am strongly against stabilizing this in `cargo clean`.  Whichever way it goes, our different stances show we have fairly fundamental conversations to have around the UX which means anything can change.  As such I've been taking the approach of "focus on the core with the UX being a mechanism to exercise the core".

Sorry, I thought I had made it clear this clear at the outset.